src/ini.cr

```diff
@@ -1,22 +1,58 @@
 class INI
+  # Exception thrown on an INI parse error.
+  class ParseException < Exception
+    getter line_number : Int32
+    getter column_number : Int32
+
+    def initialize(message, @line_number, @column_number)
+      super "#{message} at #{@line_number}:#{@column_number}"
+    end
+
+    def location
+      {line_number, column_number}
+    end
+  end
+
   # Parses INI-style configuration from the given string.
+  # Raises a `ParseException` on any errors.
   #
   # ```
   # INI.parse("[foo]\na = 1") # => {"foo" => {"a" => "1"}}
   # ```
   def self.parse(str) : Hash(String, Hash(String, String))
-    ini = {} of String => Hash(String, String)
+    ini = Hash(String, Hash(String, String)).new
+    current_section = ini[""] = Hash(String, String).new
+    lineno = 0
 
-    section = ""
     str.each_line do |line|
-      if line =~ /\s*(.*[^\s])\s*=\s*(.*[^\s])/
-        ini[section] ||= {} of String => String if section == ""
-        ini[section][$1] = $2
-      elsif line =~ /\[(.*)\]/
-        section = $1
-        ini[section] = {} of String => String
+      lineno += 1
+      next if line.empty?
+
+      offset = 0
+      line.each_char do |char|
+        break unless char.ascii_whitespace?
+        offset += 1
+      end
+
+      case line[offset]
+      when '#', ';'
+        next
+      when '['
+        end_idx = line.index(']', offset)
+        raise ParseException.new("unterminated section", lineno, line.size) unless end_idx
+        raise ParseException.new("data after section", lineno, end_idx + 1) unless end_idx == line.size - 1
+
+        current_section_name = line[offset + 1...end_idx]
+        current_section = ini[current_section_name] ||= Hash(String, String).new
+      else
+        key, eq, value = line.partition('=')
+        raise ParseException.new("expected declaration", lineno, key.size) if eq != "="
+
+        current_section[key.strip] = value.strip
       end
     end
+
+    ini.delete("") if ini[""].empty?
```

Comment:
I'm not sure this is what we want.

Reply:
It's consistent with the current parser's behavior: if there's nothing in the default/global section, it gets omitted entirely.

I can remove it, but some of the specs will need to be updated, e.g.:

```
Failure/Error: INI.parse("[section]").should eq({"section" => {} of String => String})

       Expected: {"section" => {}}
            got: {"" => {}, "section" => {}}
```

Reply:
I'm not sure about this, I still need to be convinced either way.

Reply:
I prefer the current behavior for a few reasons:

* Neither `inih` nor Python's ConfigParser includes the default section if it's empty.
* Parsing an empty string would result in `{"" => {}}` rather than just `{}`, which I personally find unintuitive.
* As above, `empty?` would be `false` on `parse("")`, which I also find unintuitive.

Reply:
Then we should delete all empty sections.

Reply:
The benefit of having empty sessions is that they're at least explicit, e.g. if I really did want an empty default session, I could do this:

```ini
[]

[foo]
bar=baz
```

But `inih` and other INI parsers don't seem to do that consistently, so I can delete all empty sessions if you think that's better.

Reply:
@woodruffw but adding `[]` doesn't make an empty default section in your code, it'd be empty and deleted regardless. But other empty sections with non-empty names wouldn't get deleted.

We shouldn't treat the default section differently is all i'm pointing out.

Reply:
Ah, you're right. I was mixing it up with the old `[]` behavior.

Reply:
It could be checked if the default section was just used implicitly or explicitly specified with `[]` and in the latter case, keep it even if empty.
But the current solution to delete all empty sections is probably fine as well.